source/FreeRTOS_IP.c

```diff
@@ -1441,84 +1441,106 @@ BaseType_t xSendEventStructToIPTask( const IPStackEvent_t * pxEvent,
-
-        if( pxEndPoint != NULL )
+        switch( pxEthernetHeader->usFrameType )
         {
-            /* The packet was directed to this node - process it. */
-            eReturn = eProcessBuffer;
-        }
-        else if( memcmp( xBroadcastMACAddress.ucBytes, pxEthernetHeader->xDestinationAddress.ucBytes, sizeof( MACAddress_t ) ) == 0 )
-        {
-            /* The packet was a broadcast - process it. */
-            eReturn = eProcessBuffer;
+            case ipARP_FRAME_TYPE:
+            case ipIPv4_FRAME_TYPE:
+                #if ipconfigIS_DISABLED( ipconfigUSE_IPv4 )
+                    eReturn = eReleaseBuffer;
+                #endif
+                break;
+
+            case ipIPv6_FRAME_TYPE:
+                #if ipconfigIS_DISABLED( ipconfigUSE_IPv6 )
+                    eReturn = eReleaseBuffer;
+                #endif
+                break;
+
+            default:
+                if( FreeRTOS_ntohs( pxEthernetHeader->usFrameType ) <= 0x600U )
+                {
+                    #if ipconfigIS_ENABLED( ipconfigFILTER_OUT_NON_ETHERNET_II_FRAMES )
+                        eReturn = eReleaseBuffer;
+                    #endif
+                }
+                else
+                {
+                    #if ipconfigIS_DISABLED( ipconfigPROCESS_CUSTOM_ETHERNET_FRAMES )
+                        eReturn = eReleaseBuffer;
+                    #endif
+                }
+                break;
         }
-        else
-        #if ( ( ipconfigUSE_LLMNR == 1 ) && ( ipconfigUSE_DNS != 0 ) )
-            if( memcmp( xLLMNR_MacAddress.ucBytes, pxEthernetHeader->xDestinationAddress.ucBytes, sizeof( MACAddress_t ) ) == 0 )
-            {
-                /* The packet is a request for LLMNR - process it. */
-                eReturn = eProcessBuffer;
-            }
-            else
-        #endif /* ipconfigUSE_LLMNR */
-        #if ( ( ipconfigUSE_MDNS == 1 ) && ( ipconfigUSE_DNS != 0 ) )
-            if( memcmp( xMDNS_MacAddress.ucBytes, pxEthernetHeader->xDestinationAddress.ucBytes, sizeof( MACAddress_t ) ) == 0 )
-            {
-                /* The packet is a request for MDNS - process it. */
-                eReturn = eProcessBuffer;
-            }
-            else
-        #endif /* ipconfigUSE_MDNS */
-        if( ( pxEthernetHeader->xDestinationAddress.ucBytes[ 0 ] == ipMULTICAST_MAC_ADDRESS_IPv6_0 ) &&
-            ( pxEthernetHeader->xDestinationAddress.ucBytes[ 1 ] == ipMULTICAST_MAC_ADDRESS_IPv6_1 ) )
+
+        if( eReturn == eReleaseBuffer )
         {
-            /* The packet is a request for LLMNR - process it. */
-            eReturn = eProcessBuffer;
+            /* No use to do more testing. */
+            break;
         }
-        else
+
+        /* Examine the destination MAC from the Ethernet header to see if it matches
+         * that of an end point managed by FreeRTOS+TCP. */
+        pxEndPoint = FreeRTOS_MatchingEndpoint( NULL, pucEthernetBuffer );
```

Comment:
This is fairly expensive search for an end-point and wasn't it already done in the network driver?

Reply:
Yes, I believe either we need to assign the endpoint to the network buffer here or just use `FreeRTOS_FindEndPointOnMAC` as a preliminary check to match the MAC address along with LLMNR and MDNS MAC addresses `if` checks.
It doesn't make much sense to ignore the endpoint result here.

Reply:
Well if `FreeRTOS_MatchingEndpoint` is called from the NetworkInterface and `ipconfigETHERNET_DRIVER_FILTERS_FRAME_TYPES == 0`, then neither should be done. If `ipconfigETHERNET_DRIVER_FILTERS_FRAME_TYPES == 1`, the previous call to `FreeRTOS_FindEndPointOnMAC` is pretty much wasted work when it is half of the work done in `FreeRTOS_MatchingEndpoint`. But you have no way of assigning the pxEndpoint result unless you change the parameter or you are able to use `pxPacketBuffer_to_NetworkBuffer`.

Reply:
I am a bit confused. We cannot "assign the endpoint" in `eConsiderFrameForProcessing()` because:

- `eConsiderFrameForProcessing()` is an optional function. It's presence is controlled by `ipconfigETHERNET_DRIVER_FILTERS_FRAME_TYPES` so if `ipconfigETHERNET_DRIVER_FILTERS_FRAME_TYPES` is enabled, the endoint needs to be assigned elswhere.
- `eConsiderFrameForProcessing()` does not have access to the network buffer, so we cannot assign an endpoint in it.

I do agree that we should use `FreeRTOS_FindEndPointOnMAC()` to check if we need to drop the frame, however since  `eConsiderFrameForProcessing()` doesn not have a pointer to the network interface in used to call:
```
pxEndPoint = FreeRTOS_FindEndPointOnMAC( &( pxEthernetHeader->xDestinationAddress ), NULL );
```
which is a bit useless as it can match the MAC of a different network interface.

My original comment was due to this:
The network driver ( or at least the DriverSAM that I'm familiar with ) already did:
```
pxNextNetworkBufferDescriptor->pxInterface = pxMyInterface;
pxNextNetworkBufferDescriptor->pxEndPoint = FreeRTOS_MatchingEndpoint( pxMyInterface, pxNextNetworkBufferDescriptor->pucEthernetBuffer );

if( pxNextNetworkBufferDescriptor->pxEndPoint == NULL )
{
    FreeRTOS_printf( ( "NetworkInterface: can not find a proper endpoint\n" ) );
    xRelease = pdTRUE;
}
```
Therefor, if we ever reached `eConsiderFrameForProcessing()`, the frame already has a good endpoint assigned.
As a matter of fact, calling
```
pxEndPoint = FreeRTOS_MatchingEndpoint( NULL, pucEthernetBuffer );
```
actually creates a problem of its own: `FreeRTOS_MatchingEndpoint()` internally will call `pxEasyFit()` with a NULL for an interface pointer. That causes `pxEasyFit()` to search on ALL interfaces and potentially return a match on IF_X even though the frame was received on IF_Y

I hope the above is clear, or maybe I'm missing something.

Reply:
@HTRamsey,
`FreeRTOS_MatchingEndpoint()` does not guarantee that the frame's destination MAC matches one of our end-point's MAC addresses. `FreeRTOS_MatchingEndpoint()` and `pxEasyFit()` could select an end-point base on IP-version or destination IP address, so with `ipconfigETHERNET_DRIVER_FILTERS_FRAME_TYPES == 0` the ONLY place that ever verifies the destination MAC of a frame against the MAC addresses of our endpoints is when `eConsiderFrameForProcessing()` does this:
```
pxEndPoint = FreeRTOS_FindEndPointOnMAC( &( pxEthernetHeader->xDestinationAddress ), NULL );

if( pxEndPoint != NULL )
{
    /* The packet was directed to this node - process it. */
    eReturn = eProcessBuffer;
}
```
I believe the code above is required, but if I were to ever try to improve `eConsiderFrameForProcessing()`, I would change it's parameter to pxNetworkBuffer and change the call above to:
```
pxEndPoint = FreeRTOS_FindEndPointOnMAC( &( pxEthernetHeader->xDestinationAddress ), pxNetworkBuffer );
```

I agree that calling `FreeRTOS_FindEndPointOnMAC()` does work that was already done in `FreeRTOS_MatchingEndpoint()` however, `FreeRTOS_MatchingEndpoint()` does not store information on whether it found a MAC match. If it could store that info somehow, then we wouldn't have to call `FreeRTOS_FindEndPointOnMAC()`

Reply:
I missed the prioritized methods in pxEasyFit, so you're right it could just check `rMATCH_IP_ADDR`. The other methods in pxEasyFit are checked in `prvAllowIPPacketIPv4` & `prvAllowIPPacketIPv6` where it calls `FreeRTOS_FindEndPointOnIP_IPv4`, `FreeRTOS_FindEndPointOnMAC`, etc. I think this is where it would be beneficial to implement what I had mentioned [here](https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/pull/804#issuecomment-1913383898). I don't know how much it affects performance, but there are a few things that are checked multiple times, like endpoints and the IP header, etc.

Reply:
Oh wow! I had not seen this. I just searched for `ipCONSIDER_FRAME_FOR_PROCESSING` in all the network drivers and you're right. It is defined all over the place. 
BTW, from my prespective, the same5x driver defines it exactly as FreeRTOS_IP.c and every other network driver that defines `ipCONSIDER_FRAME_FOR_PROCESSING` is backwards.

I don't want to speculate on what those drivers are doing. It would be nice to get some insight from someone with more long-term experience but I think this macro shouldn't be defined in the network drivers.

Reply:
I think it was intended to reduce a little overhead by dropping packets earlier, but maybe it can be placed before actually queuing the `eNetworkRxEvent` outside of the network interface. The less that's required in the network interface, the better in my opinion.

Reply:
@HTRamsey wrote:
> But you have no way of assigning the pxEndpoint result unless you change the parameter or you are able to use `pxPacketBuffer_to_NetworkBuffer`.

As discussed earlier, it is proposed that `eConsiderFrameForProcessing()` gets the packet descriptor as a parameter. 

I am not a fan of using `pxPacketBuffer_to_NetworkBuffer` unless really necessary.